drivers/adc/adc_stm32.c

```diff
@@ -0,0 +1,288 @@
+
+	void (*irq_cfg_func)(void);
+
+	struct stm32_pclken pclken;
+	struct device *p_dev;
+};
+
+static int check_buffer_size(const struct adc_sequence *sequence,
+			     u8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(u16_t);
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_ERR("Provided buffer is too small (%u/%u)",
+				sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+	return 0;
+}
+
+static void adc_stm32_start_conversion(struct device *dev)
+{
+	const struct adc_stm32_cfg *config = dev->config->config_info;
+
+	LOG_INF("Starting conversion");
+
+	adc_stm32_start_conversion_soc(config->base);
+
+	LOG_INF("Conversion done!");
+}
+
+static int start_read(struct device *dev, const struct adc_sequence *sequence)
+{
+	struct adc_stm32_data *data = dev->driver_data;
+	const struct adc_stm32_cfg *cfg = dev->config->config_info;
+
+	int err;
+
+	switch (sequence->resolution) {
+	case 6:
+		break;
+	case 8:
+		break;
+	case 10:
+		break;
+	case 12:
+		break;
+	default:
+		LOG_ERR("Invalid resolution");
+		return -EINVAL;
+	}
+
+	u32_t channels = sequence->channels;
+
+	data->buffer = sequence->buffer;
+	u8_t index;
+
+	index = find_lsb_set(channels) - 1;
+	adc_stm32_setup_channels_soc(cfg->base, index, data->channel_count);
+	adc_stm32_setup_sequence_length_soc(cfg->base, 1);
+	data->channel_count = 1;
+
+	err = check_buffer_size(sequence, data->channel_count);
+	if (err) {
+		return err;
+	}
+
+	adc_stm32_setup_resolution_soc(cfg->base, sequence->resolution);
+	adc_stm32_enable_eoc_soc(cfg->base);
+
+	adc_context_start_read(&data->ctx, sequence);
+	err = adc_context_wait_for_completion(&data->ctx);
+	adc_context_release(&data->ctx, err);
+
+	return err;
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_stm32_data *data =
+		CONTAINER_OF(ctx, struct adc_stm32_data, ctx);
+
+	data->repeat_buffer = data->buffer;
+
+	adc_stm32_start_conversion(data->dev);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_stm32_data *data =
+		CONTAINER_OF(ctx, struct adc_stm32_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static void adc_stm32_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	struct adc_stm32_data *data = (struct adc_stm32_data *)dev->driver_data;
+	struct adc_stm32_cfg *config =
+		(struct adc_stm32_cfg *)dev->config->config_info;
+	ADC_TypeDef *adc = config->base;
+
+	*data->buffer = adc->DR;
+	adc_context_on_sampling_done(&data->ctx, dev);
+
+	LOG_INF("ISR triggered.");
+}
+
+static int adc_stm32_read(struct device *dev,
+			  const struct adc_sequence *sequence)
+{
+	struct adc_stm32_data *data = dev->driver_data;
+
+	adc_context_lock(&data->ctx, false, NULL);
+
+	return start_read(dev, sequence);
```

Comment:
You need to correct the locking scheme. Currently if an invalid sequence is requested (with an invalid resolution, for instance), the driver will remain locked after reporting the error.
See for instance how it is done in [adc_nrfx_adc](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/adc/adc_nrfx_adc.c#L214) (the [recent commit](https://github.com/zephyrproject-rtos/zephyr/commit/0906a51dacc947c5fa19a80e208549f9039556cd#diff-f8d824106434c62f9e6458df9242c4f5) that modified this driver contains a similar correction to what you need to apply) or in [adc_mcux_adc16](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/adc/adc_mcux_adc16.c#L150) that looks quite similar to your driver.

There is a test for ADC drivers in the Zephyr source tree that will help you check if the API implementation is correct. See [tests/drivers/adc/adc_api](https://github.com/zephyrproject-rtos/zephyr/tree/master/tests/drivers/adc/adc_api).